src/axes.jl

```diff
@@ -0,0 +1,158 @@
+
+"""
+    axes_types(::Type{T}, dim)
+
+Returns the axis type along dimension `dim`.
+"""
+axes_types(x, dim) = axes_types(typeof(x), dim)
+@inline axes_types(::Type{T}, dim) where {T} = axes_types(T, to_dims(T, dim))
+@inline function axes_types(::Type{T}, dim::StaticInt{D}) where {T,D}
+    if D > ndims(T)
+        return OptionallyStaticUnitRange{One,One}
+    else
+        return _get_tuple(axes_types(T), dim)
+    end
+end
+@inline function axes_types(::Type{T}, dim::Int) where {T}
+    if dim > ndims(T)
+        return OptionallyStaticUnitRange{One,One}
+    else
+        return axes_types(T).parameters[dim]
+    end
+end
+
+"""
+    axes_types(::Type{T}) -> Type
+
+Returns the type of the axes for `T`
+"""
+axes_types(x) = axes_types(typeof(x))
+axes_types(::Type{T}) where {T<:Array} = Tuple{Vararg{OneTo{Int},ndims(T)}}
+function axes_types(::Type{T}) where {T}
+    if parent_type(T) <: T
+        return Tuple{Vararg{OptionallyStaticUnitRange{One,Int},ndims(T)}}
+    else
+        return axes_types(parent_type(T))
+    end
+end
+function axes_types(::Type{T}) where {T<:VecAdjTrans}
+    return Tuple{OptionallyStaticUnitRange{One,One},axes_types(parent_type(T), One())}
+end
+function axes_types(::Type{T}) where {T<:MatAdjTrans}
+    return eachop_tuple(_get_tuple, axes_types(parent_type(T)), to_parent_dims(T))
+end
+function axes_types(::Type{T}) where {T<:PermutedDimsArray}
+    return eachop_tuple(_get_tuple, axes_types(parent_type(T)), to_parent_dims(T))
+end
+function axes_types(::Type{T}) where {T<:AbstractRange}
+    if known_length(T) === nothing
+        return Tuple{OptionallyStaticUnitRange{One,Int}}
+    else
+        return Tuple{OptionallyStaticUnitRange{One,StaticInt{known_length(T)}}}
+    end
+end
+function axes_types(::Type{T}) where {N,T<:Base.ReshapedArray{<:Any,N}}
+    return Tuple{Vararg{OptionallyStaticUnitRange{One,Int},N}}
+end
+
+_int_or_static_int(::Nothing) = Int
+_int_or_static_int(x::Int) = StaticInt{x}
+
+@inline function axes_types(::Type{T}) where {N,P,I,T<:SubArray{<:Any,N,P,I}}
+    return eachop_tuple(_sub_axis_type, T, to_parent_dims(T))
+end
+@inline function _sub_axis_type(::Type{A}, dim::StaticInt) where {T,N,P,I,A<:SubArray{T,N,P,I}}
+    return OptionallyStaticUnitRange{
+        _int_or_static_int(known_first(axes_types(P, dim))),
+        _int_or_static_int(known_length(_get_tuple(I, dim)))
+    }
+end
+
+function axes_types(::Type{R}) where {T,N,S,A,R<:ReinterpretArray{T,N,S,A}}
+    if _is_reshaped(R)
+        if sizeof(S) === sizeof(T)
+            return axes_types(A)
+        elseif sizeof(S) > sizeof(T)
+            return eachop_tuple(_reshaped_axis_type, R, to_parent_dims(R))
+        else
+            return eachop_tuple(axes_types, A, to_parent_dims(R))
+        end
+    else
+        return eachop_tuple(_non_reshaped_axis_type, R, to_parent_dims(R))
+    end
+end
+
+function _reshaped_axis_type(::Type{R}, dim::StaticInt) where {T,N,S,A,R<:ReinterpretArray{T,N,S,A}}
```

Comment:
Do you recall if it was sufficient to just add the static parameter like this

```julia
function _reshaped_axis_type(::Type{R}, dim::StaticInt{D}) where {T,N,S,A,R<:ReinterpretArray{T,N,S,A},D}
```

Or do we need to call `StaticInt(D)` in between each pass?

Reply:
I don't recall, but suspect it is enough.

Reply:
We can handle that later. I'm making a PR to add support for `BitArray`s, and it'd be nice to avoid merge conflicts.
If we want to address things like `dim::StaticInt` (which should be failing `@inferred` tests if it was causing problems), we can do that in another PR.

I saw problems when these went through a chain of calls. I don't know what the heuristics are, but it's probably fine below a certain depth.